proof/crefine/RISCV64/TcbQueue_C.thy

```diff
@@ -1392,21 +1394,23 @@ lemma rf_sr_tcb_update_no_queue:
   apply (clarsimp simp: map_comp_update projectKO_opt_tcb cvariable_relation_upd_const
                         typ_heap_simps')
   apply (intro conjI)
-       subgoal by (clarsimp simp: cmap_relation_def map_comp_update projectKO_opts_defs inj_eq)
+        subgoal by (clarsimp simp: cmap_relation_def map_comp_update projectKO_opts_defs inj_eq)
+       apply (erule iffD1 [OF cmap_relation_cong, OF refl refl, rotated -1])
+       apply simp
+       apply (rule cendpoint_relation_upd_tcb_no_queues, assumption+)
+        subgoal by fastforce
+       subgoal by fastforce
       apply (erule iffD1 [OF cmap_relation_cong, OF refl refl, rotated -1])
       apply simp
-      apply (rule cendpoint_relation_upd_tcb_no_queues, assumption+)
+      apply (rule cnotification_relation_upd_tcb_no_queues, assumption+)
        subgoal by fastforce
       subgoal by fastforce
-     apply (erule iffD1 [OF cmap_relation_cong, OF refl refl, rotated -1])
-     apply simp
-     apply (rule cnotification_relation_upd_tcb_no_queues, assumption+)
+     subgoal by (clarsimp simp: map_comp_update projectKO_opt_sc typ_heap_simps'
+                                refill_buffer_relation_def)
```

Comment:
Out of curiosity, what is this `subgoal` for? Is it a long running command?

Reply:
The commented out line will break that subgoal into two. It's related to the references of `thread`. I think we don't want `thread` to be the idle thread, and it's more complicated for MCS because of the idle sc. I spent a while trying to figure it out but seeing that it's not currently mentioned in the proof code at the moment, maybe it won't be used in the end.

Reply:
yes that is fine for long(ish) running commands.

I'm not entirely sure about the indentation on this one. I think so far we mostly have `by` on the same line as `subgoal` when we don't need to break, but have it in the next line indented by 2 if things get bigger. This policy comes from Isar proofs where you would do the same fore `have` commands and similar if they are proved with `by`.